File: backend_python/src/controllers/user_controller.py
from flask import request, jsonify
from flask_jwt_extended import create_access_token, get_jwt_identity
from models.user_model import User
from config.database import db
from middleware.auth_middleware import auth_required
from werkzeug.security import generate_password_hash

def register():
    try:
        data = request.get_json()
        
        if not data:
            return jsonify({"message": "No se recibieron datos"}), 400
            
        username = data.get('username')
        email = data.get('email')
        password = data.get('password')
        profile_picture_url = data.get('profile_picture')
        
        if not username or not email or not password:
            return jsonify({"message": "Faltan datos obligatorios"}), 400
        
        existing_user = User.query.filter_by(username=username).first()
        if existing_user:
            return jsonify({"message": "El nombre de usuario ya existe"}), 409
            
        existing_email = User.query.filter_by(email=email).first()
        if existing_email:
            return jsonify({"message": "El correo electronico ya esta registrado"}), 409
        
        new_user = User(
            username=username,
            email=email,
            password=password,
            profile_picture=profile_picture_url
        )

        db.session.add(new_user)
        db.session.commit()

        return jsonify({"message": "Usuario registrado exitosamente"}), 201
    except Exception as e:
        print(f"Error en register: {str(e)}")
        return jsonify({"message": f"Error al registrar usuario: {str(e)}"}), 500
    
def login():
    try:
        # Try to get data from both form-data and JSON
        data = request.get_json() or {}
        
        username = request.form.get('username') or data.get('username')
        password = request.form.get('password') or data.get('password')
        
        if not username or not password:
            return jsonify({"message": "Datos de inicio de sesión incompletos"}), 400
            
        user = User.query.filter_by(username=username).first()
        
        if not user or not user.check_password(password):
            return jsonify({"message": "Credenciales inválidas"}), 401
        
        # Generate JWT token
        access_token = create_access_token(identity=str(user.id))
        
        return jsonify({
            "message": "Inicio de sesión exitoso",
            "token": access_token,
            "user": user.to_dict()
        }), 200
    except Exception as e:
        print("Error en login:", str(e))
        return jsonify({"message": str(e)}), 500

File: backend_nodejs/src/controllers/userController.js
const User = require('../models/userModel');
const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');

exports.register = async (req, res) => {
    try {
        const { username, email, password, profile_picture } = req.body;
        
        const existingUser = await User.findOne({ where: { username } });
        if (existingUser) {
            return res.status(400).json({ message: 'El nombre de usuario ya existe' });
        }

        const salt = await bcrypt.genSalt(10);
        const hashedPassword = await bcrypt.hash(password, salt);

        const newUser = await User.create({
            username,
            email,
            password: hashedPassword,
            profile_picture,
        });

        res.json(newUser);
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

exports.login = async (req, res) => {
    try {
        const { username, password } = req.body;
        const user = await User.findOne({ where: { username } });
        if (!user) return res.status(400).json({ message: 'Usuario no encontrado' });

        const validPassword = await bcrypt.compare(password, user.password);
        if (!validPassword) return res.status(400).json({ message: 'Contraseña incorrecta' });

        const token = jwt.sign({ id: user.id }, process.env.JWT_SECRET, { expiresIn: '1h' });
        res.json({ token, user });
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};